File: critter/core/src/main/java/dev/morphia/critter/parser/gizmo/PropertyAccessorGenerator.java
package dev.morphia.critter.parser.gizmo;

import java.util.Map;

import dev.morphia.critter.Critter;
import dev.morphia.critter.CritterClassLoader;
import dev.morphia.critter.parser.ExtensionFunctions;

import org.bson.codecs.pojo.PropertyAccessor;
import org.objectweb.asm.Type;
import org.objectweb.asm.tree.FieldNode;
import org.objectweb.asm.tree.MethodNode;

import io.quarkus.gizmo.MethodDescriptor;
import io.quarkus.gizmo.ResultHandle;

import static io.quarkus.gizmo.MethodDescriptor.ofConstructor;
import static io.quarkus.gizmo.MethodDescriptor.ofMethod;
import static io.quarkus.gizmo.SignatureBuilder.forClass;
import static io.quarkus.gizmo.SignatureBuilder.forMethod;
import static io.quarkus.gizmo.Type.classType;
import static io.quarkus.gizmo.Type.parameterizedType;
import static io.quarkus.gizmo.Type.typeVariable;
import static io.quarkus.gizmo.Type.voidType;

public class PropertyAccessorGenerator extends BaseGizmoGenerator {
    private static final Map<String, String> PRIMITIVE_TO_WRAPPER = Map.of(
            "boolean", "java.lang.Boolean",
            "byte", "java.lang.Byte",
            "char", "java.lang.Character",
            "short", "java.lang.Short",
            "int", "java.lang.Integer",
            "long", "java.lang.Long",
            "float", "java.lang.Float",
            "double", "java.lang.Double");

    private final String propertyName;
    private final String propertyType;

    public PropertyAccessorGenerator(Class<?> entity, CritterClassLoader critterClassLoader, FieldNode field) {
        super(entity, critterClassLoader);
        this.propertyName = field.name;
        this.propertyType = Type.getType(field.desc).getClassName();
        generatedType = baseName + "." + Critter.titleCase(propertyName) + "Accessor";
    }

    public PropertyAccessorGenerator(Class<?> entity, CritterClassLoader critterClassLoader, MethodNode method) {
        super(entity, critterClassLoader);
        this.propertyName = ExtensionFunctions.getterToPropertyName(method, entity);
        this.propertyType = Type.getReturnType(method.desc).getClassName();
        generatedType = baseName + "." + Critter.titleCase(propertyName) + "Accessor";
    }

    public boolean isPrimitive() {
        return PRIMITIVE_TO_WRAPPER.containsKey(propertyType);
    }

    public String getWrapperType() {
        return PRIMITIVE_TO_WRAPPER.getOrDefault(propertyType, propertyType);
    }

    public PropertyAccessorGenerator emit() {
        getBuilder().signature(
                forClass()
                        .addInterface(
                                parameterizedType(
                                        classType(PropertyAccessor.class),
                                        classType(propertyType))));

        try (var creator = getCreator()) {
            ctor();
            get();
            set();
        }

        return this;
    }

    private void get() {
        var method = getCreator().getMethodCreator(
                ofMethod(generatedType, "get", Object.class.getName(), Object.class.getName()));
        method.setSignature(
                forMethod()
                        .addTypeParameter(typeVariable("S"))
                        .setReturnType(classType(propertyType))
                        .addParameterType(typeVariable("S"))
                        .build());
        method.setParameterNames(new String[] { "model" });
        ResultHandle castModel = method.checkCast(method.getMethodParam(0), entity);
        MethodDescriptor toInvoke = ofMethod(entity, "__read" + Critter.titleCase(propertyName), propertyType);
        ResultHandle result = method.invokeVirtualMethod(toInvoke, castModel);
        ResultHandle boxed = isPrimitive() ? method.smartCast(result, getWrapperType()) : result;
        method.returnValue(boxed);
    }

    private void set() {
        var method = getCreator().getMethodCreator(
                ofMethod(generatedType, "set", "void", Object.class.getName(), Object.class.getName()));
        method.setSignature(
                forMethod()
                        .addTypeParameter(typeVariable("S"))
                        .setReturnType(voidType())
                        .addParameterType(typeVariable("S"))
                        .addParameterType(classType(propertyType))
                        .build());
        method.setParameterNames(new String[] { "model", "value" });
        ResultHandle castModel = method.checkCast(method.getMethodParam(0), entity);
        ResultHandle castValue;
        if (isPrimitive()) {
            ResultHandle boxed = method.checkCast(method.getMethodParam(1), getWrapperType());
            castValue = method.smartCast(boxed, propertyType);
        } else {
            castValue = method.checkCast(method.getMethodParam(1), propertyType);
        }
        MethodDescriptor toInvoke = ofMethod(entity, "__write" + Critter.titleCase(propertyName), "void", propertyType);
        method.invokeVirtualMethod(toInvoke, castModel, castValue);
        method.returnValue(null);
    }

    private void ctor() {
        var constructor = getCreator().getConstructorCreator(new String[0]);
        constructor.invokeSpecialMethod(ofConstructor(Object.class), constructor.getThis());
        constructor.returnVoid();
        constructor.close();
    }
}

File: critter/core/src/main/kotlin/dev/morphia/critter/parser/gizmo/PropertyAccessorGenerator.kt
package dev.morphia.critter.parser.gizmo

import dev.morphia.critter.parser.getterToPropertyName
import dev.morphia.critter.titleCase
import io.quarkus.gizmo.MethodDescriptor.ofConstructor
import io.quarkus.gizmo.MethodDescriptor.ofMethod
import io.quarkus.gizmo.ResultHandle
import io.quarkus.gizmo.SignatureBuilder.*
import io.quarkus.gizmo.Type.*
import org.bson.codecs.pojo.PropertyAccessor
import org.objectweb.asm.Type.getReturnType
import org.objectweb.asm.Type.getType
import org.objectweb.asm.tree.FieldNode
import org.objectweb.asm.tree.MethodNode

class PropertyAccessorGenerator : BaseGizmoGenerator {
    constructor(
        entity: Class<*>,
        critterClassLoader: dev.morphia.critter.CritterClassLoader,
        field: FieldNode,
    ) : super(entity, critterClassLoader) {
        propertyName = field.name
        propertyType = getType(field.desc).className
        generatedType = "${baseName}.${propertyName.titleCase()}Accessor"
    }

    constructor(
        entity: Class<*>,
        critterClassLoader: dev.morphia.critter.CritterClassLoader,
        method: MethodNode,
    ) : super(entity, critterClassLoader) {
        propertyName = method.getterToPropertyName(entity)
        propertyType = getReturnType(method.desc).className
        generatedType = "${baseName}.${propertyName.titleCase()}Accessor"
    }

    val propertyName: String
    val propertyType: String
    val isPrimitive: Boolean
        get() = primitiveToWrapper.containsKey(propertyType)

    val wrapperType: String
        get() = primitiveToWrapper[propertyType] ?: propertyType

    companion object {
        private val primitiveToWrapper =
            mapOf(
                "boolean" to "java.lang.Boolean",
                "byte" to "java.lang.Byte",
                "char" to "java.lang.Character",
                "short" to "java.lang.Short",
                "int" to "java.lang.Integer",
                "long" to "java.lang.Long",
                "float" to "java.lang.Float",
                "double" to "java.lang.Double",
            )
    }

    fun emit(): PropertyAccessorGenerator {
        builder.signature(
            forClass()
                .addInterface(
                    parameterizedType(
                        classType(PropertyAccessor::class.java),
                        classType(propertyType),
                    )
                )
        )

        creator.use {
            ctor()
            get()
            set()
        }

        return this
    }

    private fun get() {
        val method =
            creator.getMethodCreator(
                ofMethod(generatedType, "get", Object::class.java.name, Object::class.java.name)
            )
        method.signature =
            forMethod()
                .addTypeParameter(typeVariable("S"))
                .setReturnType(classType(propertyType))
                .addParameterType(typeVariable("S"))
                .build()
        method.setParameterNames(arrayOf("model"))
        val castModel = method.checkCast(method.getMethodParam(0), entity)
        val toInvoke = ofMethod(entity, "__read${propertyName.titleCase()}", propertyType)
        val result = method.invokeVirtualMethod(toInvoke, castModel)
        val boxed: ResultHandle = if (isPrimitive) method.smartCast(result, wrapperType) else result
        method.returnValue(boxed)
    }

    private fun set() {
        val method =
            creator.getMethodCreator(
                ofMethod(
                    generatedType,
                    "set",
                    "void",
                    Object::class.java.name,
                    Object::class.java.name,
                )
            )
        method.signature =
            forMethod()
                .addTypeParameter(typeVariable("S"))
                .setReturnType(voidType())
                .addParameterType(typeVariable("S"))
                .addParameterType(classType(propertyType))
                .build()
        method.setParameterNames(arrayOf("model", "value"))
        val castModel = method.checkCast(method.getMethodParam(0), entity)
        val castValue: ResultHandle =
            if (isPrimitive) {
                val boxed = method.checkCast(method.getMethodParam(1), wrapperType)
                method.smartCast(boxed, propertyType)
            } else method.checkCast(method.getMethodParam(1), propertyType)
        val toInvoke = ofMethod(entity, "__write${propertyName.titleCase()}", "void", propertyType)
        method.invokeVirtualMethod(toInvoke, castModel, castValue)
        method.returnValue(null)
    }

    private fun ctor() {
        val constructor = creator.getConstructorCreator(*arrayOf<String>())
        constructor.invokeSpecialMethod(ofConstructor(Object::class.java), constructor.`this`)
        constructor.returnVoid()

        constructor.close()
    }
}
